### ISSUE TITLE
use per-instance pragmas OVERLAPPING/OVERLAPPABLE/OVERLAPS

### ISSUE THREAD

**Issue by tolysz** [commercialhaskell/stack#1569]
Just playing with the `git` version of `aeson` and they started using:
`https://github.com/bos/aeson/blob/master/include/overlapping-compat.h`
It would be cool if we could mark instances
 `{-# OVERLAPPING #-}` or use the trick from above.

**Comment by mgsloan** [comment]
AFAIK we have no uses of `OverlappingInstances`.  Does the git version of aeson cause it to be necessary?

While the module-level `OverlappingInstances` is deprecated, I think since stack isn't used as a library much (if at all), we may as well use it for easy compatibility with both 7.8 and 7.10.  I /think/ these finer grained declarations are only beneficial when you don't know what other instances are going to be declared.

**Comment by tolysz** [comment]
The current `aeson-10` is a bit broken, i.e. `Maybe a` fails, however there is a PR which fixes it.
I was testing if I can compile stack against git version +PR ( https://github.com/tolysz/stack/blob/git-aeson/stack-10.yaml ) and discovered the pragmas will be needed.
And we will need this version ASAP :) but I guess it will be after new year.

I had to modify:

        modified:   src/Stack/Types/Config.hs
        modified:   src/Stack/Types/Docker.hs
        modified:   src/Stack/Types/Image.hs
        modified:   src/Stack/Types/Nix.hs

**Comment by borsboom** [comment]
Are the pragmas needed for aeson itself, or is it because of aeson-compat?  When I tried it a few weeks ago, Stack compiled against aeson-0.10 and all the tests passed, without needing any extra pragmas.  Are there any cases you've found where Stack fails with aeson-0.10?

**Comment by tolysz** [comment]
It compiles fine against `hackage`0-10 version the `git` one fails. 
It was introduced around 11 Oct https://github.com/bos/aeson/commit/8e1dbc7e723c7e20758368559f7093f1cfd5f742
The hackage version is from September.

I refer to `aeson-compat` in the `yaml` only because I fixed bug in `aeson` so I removed the workaround there. But it is not related to it. It complains about having instances for `(a,b)` already defined in `aeson` and stops.
